engine/runtime/src/main/java/org/enso/interpreter/node/typecheck/AbstractTypeCheckNode.java

```diff
@@ -0,0 +1,93 @@
+package org.enso.interpreter.node.typecheck;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.Node;
+import java.util.List;
+import org.enso.interpreter.node.ExpressionNode;
+import org.enso.interpreter.node.expression.builtin.meta.AtomWithAHoleNode;
+import org.enso.interpreter.runtime.data.text.Text;
+import org.enso.interpreter.runtime.error.DataflowError;
+
+/**
+ * Root of hierarchy of nodes checking types. This class (and its subclasses) are an implementation
+ * detail. The API to perform the is in {@link TypeCheckNode}.
+ */
+abstract sealed class AbstractTypeCheckNode extends Node
+    permits OneOfTypesCheckNode, AllOfTypesCheckNode, SingleTypeCheckNode, MetaTypeCheckNode {
+  private final String comment;
+  @CompilerDirectives.CompilationFinal private String expectedTypeMessage;
+
+  AbstractTypeCheckNode(String comment) {
+    this.comment = comment;
+  }
+
+  abstract Object findDirectMatch(VirtualFrame frame, Object value);
```

Comment:
Even though these methods are not part of the public API, I would love to see some docs for them. It is really hard for me to deduce its contracts and meaning of arguments and return values just by reading the code.

Reply:
I will need to touch these methods again while working on #11482 - I'll remember to document them.